openpype/modules/kitsu/plugins/publish/integrate_kitsu_note.py

```diff
@@ -20,16 +22,54 @@ def process(self, context):
 
         self.log.debug("Comment is `{}`".format(publish_comment))
 
+        families = set(
+            instance.data.get("kitsu_task")
+            for instance in context
+            if instance.data.get("publish")
+        )
+
         for instance in context:
             kitsu_task = instance.data.get("kitsu_task")
-            if kitsu_task is None:
+            if not kitsu_task:
                 continue
 
             # Get note status, by default uses the task status for the note
             # if it is not specified in the configuration
-            note_status = kitsu_task["task_status"]["id"]
-
-            if self.set_status_note:
+            kitsu_task = context.data["kitsu_task"]
+            shortname = kitsu_task["task_status"]["short_name"].upper()
+            note_status = kitsu_task["task_status_id"]
+            if (
+                self.set_status_note
+                and next(
+                    (
+                        False
+                        for status_except in self.status_exceptions
+                        if shortname == status_except["short_name"].upper()
+                        and status_except["condition"] == "equal"
+                        or shortname != status_except["short_name"].upper()
+                        and status_except["condition"] == "not_equal"
+                    ),
+                    True,
+                )
```

Comment:
I'd turn it into positive way and simplify it like this:
```suggestion
                and next(
                    (
                        (status_except["condition"] == "equal"
                        and shortname == status_except["short_name"].upper())
                        and (status_except["condition"] == "not_equal"
                        and shortname != status_except["short_name"].upper())
                        for status_except in self.status_exceptions
                    ),
                    False,
                )
```

Reply:
Looking at this logic I'd say it'd actually be more readable in a for loop - no?

```python
# Must not match a status exception
allow_status = False
for status_req in self.status_exceptions:
     equals = status_req ["condition"] == "equal"
     match = status_req ["short_name"].upper() == shortname
     if match == condition:
         allow_status = True
         break

# Must match a family requirement
allow_family = False
for family_req in self.family_requirements:
     equals = family_req["condition"] == "equal"
     match = family_req["short_name"].lower() == family
     if match == condition:
         allow_family = True
         break

if (
    self.status_note and
    allow_status and
    allow_family
):
    ...
```

Consider this psuedocode because I just can't easily within a minute grasp what the original if statements here are trying to do. So likely my logic is off, but in essence that'd be close to what the next() branches, etc. are doing. Right?

Reply:
Hey, your proposal is slightly more readable, but slower to execute. I always put speed first when I have the choice, but that's maybe not in the guideline.

You're right about the `lower()` though, it must be `upper()`.

Reply:
The speed difference I think in this case will be more than neglible. In your case you're performing `.lower()` call twice and the lookup into the dict of `family_req` twice too.

---

By the way, I would love some references on `next` being that much faster in comparison if you have them? Isn't that just about it using iterators and breaking early which my code basically does the same as the iterator comprehension (or the `generator` that it creates). I thought it was more about don't use list comprehensions which pre-build full lists, etc.

Reply:
I do find Tilix proposition better than what I did, except it needs to be linted and there's a mistake in his code, I think it should look more like this:
```
                and next(
                    (
                        (
                            status_except["condition"] == "equal"
                            and shortname == status_except["short_name"].upper()
                        )
                        or (
                            status_except["condition"] == "not_equal"
                            and shortname != status_except["short_name"].upper()
                        )
                        for status_except in self.status_exceptions
                    ),
                    False,
                )
```

Reply:
In fact, my proposal implies what I was suggesting here https://github.com/ynput/OpenPype/pull/4537#pullrequestreview-1343248632, when I propose to turn it in a positive way.

Also, what I propose in the current discussion is wrong if we want to match any listed status condition when the original proposal, with the `if` in the generator to actually hold a boolean if matched status, was better. Mine creates a boolean for each condition, even though they don't have anything to do with the current status, which is wrong because if the first one is `False`, we'll miss other matching conditions. For this use case we'd be tempted to use a `any()`.

About performances, here I've made a test script: [mytests.zip](https://github.com/ynput/OpenPype/files/10999239/mytests.zip)

Conclusions:
- `for` with `break` is the fastest,  @BigRoy's wins :tada: 
- `next()` is a very bit slower, probably due to the function's overhead
- `any` is far slower

Therefore, I'd recommend to use @BigRoy's proposal, in a positive way (see my comment)!